compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -1483,20 +1484,31 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
             Option<ObligationCause<'tcx>>,
         )>,
     ) -> ProbeResult {
-        debug!("consider_probe: self_ty={:?} probe={:?}", self_ty, probe);
-
         self.probe(|_| {
             // First check that the self type can be related.
-            let sub_obligations = match self.at(&ObligationCause::dummy(), self.param_env).sup(
-                DefineOpaqueTypes::Yes,
-                probe.xform_self_ty,
-                self_ty,
-            ) {
-                Ok(InferOk { obligations, value: () }) => obligations,
-                Err(err) => {
-                    debug!("--> cannot relate self-types {:?}", err);
-                    return ProbeResult::NoMatch;
+            let sub_obligations = match self_ty.kind() {
+                // HACK: opaque types will match anything for which their bounds hold.
+                // Thus we need to prevent them from trying to match the `&_` autoref
+                // candidates that get created for `&self` trait methods.
+                ty::Alias(ty::Opaque, alias_ty)
+                    if self.infcx.can_define_opaque_ty(alias_ty.def_id) =>
+                {
+                    if !probe.xform_self_ty.is_ty_var() {
+                        return ProbeResult::NoMatch;
+                    }
+                    vec![]
                 }
```

Comment:
This is only an issue for opaque types themselves, not if they are nested in anything. This is because

* references get autodereffed and we start out with the most dereffed version
* aggregates that have opaques in their associated types are either dereffable or will only get matched against infer vars or references to infer vars